### PR TITLE
fix for perf issue during fqdn resolution

### PR DESCRIPTION
**Pull request by bpjain2004**
# Description
fix for perf issue during fqdn resolution
The existing implementation tries to update the client fqdn as a client in the export. If the fqdn is not resolvable, it tries the client IP before giving up. This is repeated for every controller publish call. Recently it is observed that sometimes the Isilon takes nearly 90 seconds to return with error to the clients if fqdn is not resolvable. The fix is to avoid this delay where the driver remembers the successful client(fqdn or ip) used to populate the export during the first call and to use the same in future calls.

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
https://github.com/dell/csm/issues/190

# Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [ ] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] Backward compatibility is not broken

# How Has This Been Tested?
- tested with a secret with multiple clusters
- tested on multi-node setup
- have run tests to measure perf improvements

### CODE REVIEW

**Comment by randeepsharma** [comment]
Need to link this to a github csm issue. Also please add a bit more information in the description.

**Comment by bpjain2004** [comment]
> Need to link this to a github csm issue. Also please add a bit more information in the description.

done

**Review by prablr79 — APPROVED** [review]

File: service/isiService.go [hunk 1]
```diff
@@ -464,9 +463,73 @@ func (svc *isiService) OtherClientsAlreadyAdded(ctx context.Context, exportID in
 	return clientFieldsNotEmpty && !isNodeInClientFields && !isNodeFQDNInClientFields
 }
 
-func (svc *isiService) AddExportClientNetworkIdentifierByIDWithZone(ctx context.Context, exportID int, accessZone, nodeID string, addClientFunc func(ctx context.Context, exportID int, accessZone, clientIP string) error) error {
+// updateClusterToNodeIDMap updates cluster to nodeID map from input clusterName, nodeID and clientToUse
+func updateClusterToNodeIDMap(ctx context.Context, clusterToNodeIDMap *sync.Map, clusterName, nodeID, clientToUse string) error {
```
Comment (randeepsharma) [inline service/isiService.go:467]: our files are getting long, since these are not related to service calls, we can have a different file for the newly added functions.

File: service/isiService.go [hunk 2]
```diff
@@ -477,21 +540,41 @@ func (svc *isiService) AddExportClientNetworkIdentifierByIDWithZone(ctx context.
 		return err
 	}
 
-	log.Debugf("AddExportClientNetworkIdentifierByID client FQDN '%s' client IP '%s'", clientFQDN, clientIP)
+	currentClient, err := getClientToUseForNodeID(ctx, clusterToNodeIDMap, clusterName, nodeID)
```
Comment (randeepsharma) [inline service/isiService.go:543]: since the clusterToNodeIDMap is package variable, we need not pass it. not necessary to change it.

File: service/isiService.go (hunk 1 above)
Comment (bpjain2004) [reply service/isiService.go:467]: though these are not CSI RPC calls, these are helper functions. These are not even util functions that can be moved to utils package- these functions work on sync map, and it is not a good idea to have functions operate on this from another package.
Agree that we can have a new file for service related helper functions/methods, we can take up that as part of a separate ticket. There are many more such functions present in service package.

File: service/isiService.go [hunk 3]
```diff
@@ -464,9 +463,73 @@ func (svc *isiService) OtherClientsAlreadyAdded(ctx context.Context, exportID in
...
+}
+
+// getClientToUseForNodeID returns client to use for an input nodeID from the cluster to nodeID map, if present.
+// Otherwise returns an error
+func getClientToUseForNodeID(ctx context.Context, clusterToNodeIDMap *sync.Map, clusterName, nodeID string) (string, error) {
+	log := utils.GetRunIDLogger(ctx)
+
+	var nodeIDToClientMaps []*nodeIDToClientMap
+
+	if m, found := clusterToNodeIDMap.Load(clusterName); found {
```
[the remark this answers is not in the document]
Comment (bpjain2004) [reply service/isiService.go:511]: this is a local loop variable for map and in golang it is usually this notion is followed as long as there is no ambiguity.

File: service/isiService.go (hunk 2 above)
Comment (bpjain2004) [reply service/isiService.go:543]: agree with you, but I went for this signature considering making this available for other maps also if we introduce them in the future. Otherwise, it will be tied to only this variable.

Comment (randeepsharma) [reply service/isiService.go:543]: Okay.

File: service/isiService.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (randeepsharma) [reply service/isiService.go:511]: ok. no worries.

**Review by randeepsharma — APPROVED** [review]
